### ISSUE TITLE
Memory leak in LLVM causing asan builds to fail

### ISSUE THREAD

**Issue by SanderVocke** [halide/Halide#3090]
When building projects with clang Address Sanitizer, Halide generators fail because asan detects a memory leak in `RegisterHandlers()`. Since the generator run return code becomes nonzero due to this leak, the generation step is seen as failed and the project is unable to build.

The following is a basic project which can be used to trigger the issue if built with clang and with HALIDE_DISTRIB_DIR set. It should also be easy to reproduce by building any project using generators with clang and asan enabled.

CMakeLists.txt:
```
cmake_minimum_required(VERSION 3.1.0)

set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsanitize=address")

include("${HALIDE_DISTRIB_DIR}/halide.cmake")

halide_library(
	my_generator
	SRCS gen.cpp
	)
    target_link_libraries(my_generator.generator_binary PRIVATE ${HALIDE_LIBRARIES} ${HALIDE_SYSTEM_LIBS})
    target_include_directories(my_generator.generator_binary SYSTEM PRIVATE ${HALIDE_INCLUDE_DIR})


add_executable(my_app app.cpp)
target_link_libraries(my_app my_generator)
```

gen.cpp:
```
#include <Halide.h>

namespace {

using namespace Halide;

// Generator
class MyGen : public Generator<MyGen> {
    Output<Buffer<int>> f{"f", 2};
public:

    void generate() {
        Var x,y;
        f(x,y) = x+y;
    }
};

}

HALIDE_REGISTER_GENERATOR(MyGen, my_generator)
```

app.cpp:
```
int main() {
    return 0;
}
```

### Result

```
FAILED: genfiles/halide_rt_host/halide_rt_host.a 
cd /home/sander/projects/halide_bug_asan/build-clang-asan-Debug && /usr/bin/cmake -E echo Running /home/sander/projects/halide_bug_asan/build-clang-asan-Debug/halide_library_runtime.generator_binary -o /home/sander/projects/halide_bug_asan/build-clang-asan-Debug/./genfiles/halide_rt_host -r halide_rt_host target=host && /home/sander/projects/halide_bug_asan/build-clang-asan-Debug/halide_library_runtime.generator_binary -o /home/sander/projects/halide_bug_asan/build-clang-asan-Debug/./genfiles/halide_rt_host -r halide_rt_host target=host
Running /home/sander/projects/halide_bug_asan/build-clang-asan-Debug/halide_library_runtime.generator_binary -o /home/sander/projects/halide_bug_asan/build-clang-asan-Debug/./genfiles/halide_rt_host -r halide_rt_host target=host

=================================================================
==14475==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 67584 byte(s) in 1 object(s) allocated from:
    #0 0x4d3ee8 in __interceptor_malloc (/home/sander/projects/halide_bug_asan/build-clang-asan-Debug/halide_library_runtime.generator_binary+0x4d3ee8)
    #1 0x7fbf1278103b in RegisterHandlers() (/opt/halide/bin/libHalide.so+0x23d803b)

SUMMARY: AddressSanitizer: 67584 byte(s) leaked in 1 allocation(s).
```

**Comment by steven-johnson** [comment]
These "leaks" are not fixable as the entire point is to allocate a structure that is never freed. Is there a way we can add annotations to have ASAN ignore these?

**Comment by SanderVocke** [comment]
One possible workaround is an env variable setting that can be used during generator execution: set ASAN_OPTIONS to "detect_leaks=0".
This could be added to `halide.cmake` for the generator run commands. Not sure if it works for gcc address sanitizer though.

**Comment by SanderVocke** [comment]
Or better: it is possible to tell Leak Sanitizer to ignore particular leaks through a suppressions file, as explained here:
https://github.com/google/sanitizers/wiki/AddressSanitizerLeakSanitizer

**Comment by SanderVocke** [comment]
I am curious though: what is the advantage of allocating a structure that is never freed?

**Comment by steven-johnson** [comment]
It's not that there's an *advantage* to never freeing it, it's just that it's not at all important in real-world usage, since the amount allocated is trivial and fixed. I'm perfectly happy to have it fixed via either some static dtor or via annotation. (But, I'm essentially on vacation all this week and can't really look at it till next week at the earliest; would be happy to have any reasonable fix submitted as a patch of course.)

**Comment by SanderVocke** [comment]
The leak seems to occur in LLVM, not in Halide source code - looks like the generator registry already frees everything nicely. So I will have a look at working around this using an annotation instead of fixing the root cause.

**Comment by alexreinking** [comment]
Fixed by #3118
